### ISSUE TITLE
Tests fail on i386 architecture

### ISSUE THREAD

**Issue by iliastsi** [haskell/vector#186]
This causes *vector* to fail to build on Debian. You can find the build logs [here](https://buildd.debian.org/status/fetch.php?pkg=haskell-vector&arch=i386&ver=0.12.0.1-1&stamp=1505396428&raw=0). Could you please have a look at this?

**Comment by cartazio** [comment]
If you build vector with the ghc flag -sse2 this failure should probably go
away.  It would be helpful to mention the error.  That Log is long

Looks like examples where the precise internal rep and rounding matters.
Could you try that additional flag ?

You probably don’t want to use x87 floating point registers anyways.

**Comment by iliastsi** [comment]
Thank you for your quick reply!

The tests failing are:
```
Data.Vector.Primitive.Vector (Double):
  enumFromN: [Failed]
*** Failed! Falsifiable (after 60 tests and 4 shrinks): 
3.9087391205479873
30
(used seed 648261397)
  enumFromThenN: [Failed]
*** Failed! Falsifiable (after 8 tests and 2 shrinks): 
9.014121626535944
13.151835913182065
4
(used seed -1813809749)
  enumFromTo: [Failed]
*** Failed! Falsifiable (after 2 tests): 
0.8926094212244898
83.59698885055921
(used seed -183568215)

Data.Vector.Storable.Vector (Double):
  enumFromN: [Failed]
*** Failed! Falsifiable (after 91 tests and 5 shrinks): 
0.31341336491390687
9
(used seed -519395688)
  enumFromThenN: [Failed]
*** Failed! Falsifiable (after 11 tests and 4 shrinks): 
2.0
6.4130275923619555
3
(used seed 1032255777)
  enumFromTo: [Failed]
*** Failed! Falsifiable (after 2 tests): 
0.13415862562145786
46.455801269448145
(used seed 41173730)

Data.Vector.Unboxed.Vector (Double):
  enumFromThenN: [Failed]
*** Failed! Falsifiable (after 5 tests and 2 shrinks): 
0.6425150692637146
14.083850706597834
3
(used seed 581512336)
  enumFromTo: [Failed]
*** Failed! Falsifiable (after 3 tests): 
2.5290697866130847
94.11888942235991
(used seed -1565613753)
```

I tried with the -msse2 ghc flag and it worked! This is the patch that I used:
```diff
diff --git a/vector.cabal b/vector.cabal                                                                                                                                                                    
index 326454c..7e34029 100644
--- a/vector.cabal
+++ b/vector.cabal
@@ -153,6 +153,9 @@ Library

   Ghc-Options: -O2 -Wall

+  if arch(i386)
+    Ghc-Options: -msse2
+
   if !flag(Wall)
     Ghc-Options: -fno-warn-orphans

@@ -206,6 +209,9 @@ test-suite vector-tests-O0
   Ghc-Options: -O0
   Ghc-Options: -Wall

+  if arch(i386)
+    Ghc-Options: -msse2
+
   if !flag(Wall)
     Ghc-Options: -fno-warn-orphans -fno-warn-missing-signatures
     if impl(ghc >= 8.0) && impl( ghc < 8.1)
@@ -243,6 +249,9 @@ test-suite vector-tests-O2

   Ghc-Options: -O2 -Wall

+  if arch(i386)
+    Ghc-Options: -msse2
+
   if !flag(Wall)
     Ghc-Options: -fno-warn-orphans -fno-warn-missing-signatures
     if impl(ghc >= 8.0) && impl(ghc < 8.1)
```

Would you be willing to apply something like this?

**Comment by cartazio** [comment]
I think we probably should ... if only because the nixos folks hit the same
issue a year or two ago. And msse2 isn’t the default on 32bit ghc yet even
though it should be.

I’m glad to hear my guess worked out 😎🚀

**Comment by cartazio** [comment]
fixed in ghc 8.10 onwards
